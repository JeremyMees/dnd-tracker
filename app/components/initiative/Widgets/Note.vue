<script setup lang="ts">
const emit = defineEmits<{ update: [string] }>()

const props = defineProps<{ value: string }>()

const text = ref<string>(props.value)
const debouncedText = refDebounced(text, 500, { maxWait: 1000 })

watch(
  () => props.value,
  (newValue) => {
    if (newValue !== text.value) text.value = newValue || ''
  },
)

watch(debouncedText, () => emit('update', debouncedText.value))
</script>

<template>
  <TextEditor
    color="secondary"
    :content="text"
    @updated="text = $event"
  />
</template>
