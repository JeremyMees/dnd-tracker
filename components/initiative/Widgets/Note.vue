<script setup lang="ts">
const emit = defineEmits<{ update: [string] }>()

const props = defineProps<{ value: string }>()

const text = ref<string>(props.value)
const debouncedText = refDebounced(text, 500, { maxWait: 1000 })

watch(debouncedText, () => emit('update', debouncedText.value))
</script>

<template>
  <TextEditor
    color="secondary"
    :content="text"
    @updated="text = $event"
  />
</template>
