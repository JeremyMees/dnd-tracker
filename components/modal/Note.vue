<script setup lang="ts">
import { FormKitMessages } from '@formkit/vue'
import { reset } from '@formkit/core'

const emit = defineEmits<{ close: [] }>()

const props = defineProps<{
  campaignId: number
  note?: NoteRow
}>()

const input = ref()
const hiddenText = ref()
const text = ref<string>(props.note?.text || '')

const { mutateAsync: createNote } = useNoteCreate()
const { mutateAsync: updateNote } = useNoteUpdate()

onMounted(() => {
  if (input.value) focusInput(input.value)
})

async function handleSubmit(form: NoteForm, node: FormNode): Promise<void> {
  node.clearErrors()

  const onError = (err: string) => {
    reset('Note')
    node.setErrors(err)
  }

  const formData = {
    ...sanitizeForm<NoteForm>(form),
    campaign: props.campaignId,
  }

  if (props.note) {
    await updateNote({
      data: formData,
      id: props.note.id,
      onSuccess: () => emit('close'),
      onError,
    })
  }
  else {
    await createNote({
      data: formData,
      onSuccess: () => emit('close'),
      onError,
    })
  }
}
</script>

<template>
  <FormKit
    id="Note"
    type="form"
    :actions="false"
    @submit="handleSubmit"
  >
    <FormKit
      ref="input"
      name="title"
      :label="$t('components.inputs.titleLabel')"
      :value="note?.title"
      validation="required|length:3,30"
    />
    <FormKit
      ref="hiddenText"
      v-model="text"
      type="hidden"
      name="text"
      validation="required|length:10,5000"
    />

    <TextEditor
      color="background"
      :content="text"
      @updated="text = $event"
    >
      <template #error>
        <FormKitMessages :node="hiddenText?.node" />
      </template>
    </TextEditor>
  </FormKit>
</template>
