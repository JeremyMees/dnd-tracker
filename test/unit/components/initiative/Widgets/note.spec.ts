import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it } from 'vitest'
import Note from '~/components/initiative/Widgets/Note.vue'

interface Props { value: string }

const props: Props = { value: 'Test note' }

describe('Initiative note widget', async () => {
  it('Should match snapshot', async () => {
    const component = await mountSuspended(Note, { props })

    expect(component.html()).toMatchSnapshot()
  })

  it('Should emit update after debounce when text changes', async () => {
    const component = await mountSuspended(Note, { props })
    const textEditor = component.findComponent({ name: 'TextEditor' })

    // Simulate text update
    await textEditor.vm.$emit('updated', 'New note content')

    // Wait for debounce
    await new Promise(resolve => setTimeout(resolve, 600))

    expect(component.emitted('update')).toBeTruthy()
    expect(component.emitted('update')?.[0]).toEqual(['New note content'])
  })

  it('Should initialize with provided value', async () => {
    const component = await mountSuspended(Note, { props })
    const textEditor = component.findComponent({ name: 'TextEditor' })

    expect(textEditor.props('content')).toBe('Test note')
  })
})
