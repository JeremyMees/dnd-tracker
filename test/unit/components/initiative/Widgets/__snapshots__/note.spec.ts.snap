// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Initiative note widget > Should match snapshot 1`] = `
"<div class="transition-colors duration-300 border-4 rounded-lg border-secondary bg-secondary/50 py-2 px-0">
  <div class="border-b pb-1 px-2 mb-2 flex items-center gap-x-2 flex-wrap border-secondary">
    <div class="relative group w-fit"><button class="group-hover:bg-secondary px-1 pt-1 rounded-lg rounded-b-none flex flex-row items-center cursor-pointer duration-200 ease-in-out transition-all" aria-label="Text style dropdown" type="button"><span class="duration-200 font-bold text-sm min-w-10">general.text</span><span class="iconify i-tabler:chevron-down size-4" aria-hidden="true"></span></button>
      <!---->
    </div>
    <div class="flex items-center border-r pr-2 border-secondary"><button type="button" aria-label="general.link" class="icon-btn-primary size-6 text-foreground"><span class="iconify i-tabler:link size-4" aria-hidden="true"></span></button><button type="button" aria-label="actions.unlink" class="icon-btn-primary size-6 text-foreground" disabled=""><span class="iconify i-tabler:unlink size-4" aria-hidden="true"></span></button><button type="button" aria-label="general.unOrderedList" class="icon-btn-primary size-6 text-foreground"><span class="iconify i-tabler:list size-4" aria-hidden="true"></span></button><button type="button" aria-label="general.orderedList" class="icon-btn-primary size-6 text-foreground"><span class="iconify i-tabler:list-numbers size-4" aria-hidden="true"></span></button><button type="button" aria-label="general.quote" class="icon-btn-primary size-6 text-foreground"><span class="iconify i-tabler:blockquote size-4" aria-hidden="true"></span></button><button type="button" aria-label="general.horizontalRule" class="icon-btn-primary size-6 text-foreground"><span class="iconify i-tabler:separator size-4" aria-hidden="true"></span></button></div>
    <div class="flex items-center border-r pr-2 border-secondary"><button type="button" aria-label="general.bold" class="icon-btn-primary size-6 text-foreground"><span class="iconify i-tabler:bold size-4" aria-hidden="true"></span></button><button type="button" aria-label="general.italic" class="icon-btn-primary size-6 text-foreground"><span class="iconify i-tabler:italic size-4" aria-hidden="true"></span></button><button type="button" aria-label="general.strikeThrough" class="icon-btn-primary size-6 text-foreground"><span class="iconify i-tabler:strikethrough size-4" aria-hidden="true"></span></button><button type="button" aria-label="actions.highlight" class="icon-btn-primary size-6 text-foreground"><span class="iconify i-tabler:highlight size-4" aria-hidden="true"></span></button><button type="button" aria-label="actions.clearFormatting" class="icon-btn-primary size-6 text-foreground"><span class="iconify i-tabler:clear-formatting size-4" aria-hidden="true"></span></button></div>
    <div class="flex items-center"><button type="button" aria-label="actions.undo" disabled="" class="icon-btn-primary size-6 text-foreground"><span class="iconify i-tabler:arrow-back size-4" aria-hidden="true"></span></button><button type="button" aria-label="actions.redo" disabled="" class="icon-btn-primary size-6 text-foreground"><span class="iconify i-tabler:arrow-forward size-4" aria-hidden="true"></span></button></div>
  </div>
  <div class="html-richtext">
    <div contenteditable="true" spellcheck="false" translate="no" class="tiptap ProseMirror" tabindex="0">
      <p>Test note</p>
    </div>
  </div>
  <div class="flex items-center gap-2 px-2 pt-4"><svg class="size-7 -rotate-90" viewBox="0 0 36 36">
      <circle cx="18" cy="18" r="16" fill="none" stroke-width="4" class="stroke-current text-secondary"></circle>
      <circle data-test-dial="" cx="18" cy="18" r="16" fill="none" stroke-width="4" stroke-dasharray="100" stroke-dashoffset="-100" stroke-linecap="round" class="text-primary stroke-current duration-300 transition-all"></circle>
    </svg>
    <div class="flex flex-col text-xs text-muted-foreground duration-300 transition-colors"><span class="">9 / 5000 general.character</span><span>2 general.word</span></div>
  </div>
</div>"
`;
